### PR TITLE
Implementation in Rust with std thread library and message passing.

### PR DESCRIPTION
**Pull request by max-b**
The threaded implementation in rust looks similar to go in that there's a many to one message channel. I have a feeling that it's much much less efficient to do it this way because I'm spawning a new system thread per entry in a directory and they're not doing that much work each, so the overhead for spawning/managing the threads is probably outweighing any parallelism benefits. On a larger/more realistic task this approach might make more sense.

### CODE REVIEW

**Comment by max-b** [comment]
Oh - I also ended up checking the `./bin` folder into git, but ignoring everything in it. That way I didn't have to do an awkward `mkdir` inside the makefile after building the binary. Dunno if that makes sense, but let me know...

**Comment by nicolashahn** [comment]
Maybe we should generate (or make a script that generates) a bunch of larger files so that the differences can be seen more clearly? Dozens of files that are megabytes each with thousands of words so spawning threads is worth it. I could whip up a python script for this

File: Makefile [hunk 1]
```diff
@@ -1,13 +1,16 @@
...
 
 HASKELL_SRC = ./wc-haskell/src/Lib.hs ./wc-haskell/app/Main.hs
 HASKELL_BIN = ./bin/wc-haskell
 
 .PHONY: all
-all: go haskell
+all: go haskell rust
 
-.PHONY: go
+.PHONY: go rust
```
Comment (dan-f) [inline Makefile]: Nitpick, but could you please move this `.PHONY: rust` line down to above the `rust` target on line 27 for consistency?

File: wc-rust/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            output.push(dir.path());
+        }
+    }
+    Ok(output)
+}
+
+fn main() -> std::io::Result<()> {
+    let args: Vec<String> = env::args().collect();
+
+    let path = match args.len() {
```
Comment (dan-f) [inline wc-rust/src/main.rs]: Ah `match` makes this so pretty :')

File: wc-rust/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        }
+    };
+
+    let (tx, rx) = mpsc::channel();
+
+    let path_count = paths.len();
+
+    for path in paths {
+        let tx1 = mpsc::Sender::clone(&tx);
+        let _join_handle = thread::spawn(move || {
```
Comment (dan-f) [inline wc-rust/src/main.rs]: Nitpick - looks like you could remove this `_join_handle`.

**Comment by dan-f** [comment]
Beautiful! Just a couple of nitpick-y points.

**Comment by max-b** [comment]
So doing it the "naive" way of just spawning a new thread for every single file to open was nagging at me because it was just *wrong*. So I followed the last chapter in the rust book which basically walks you through creating a thread pool. There are definitely libraries (`rayon`, `threadpool`) which will handle these things for you, but I figured I might as well get the experience of walking through it. The nice thing is that the `main` code is almost identical, just passing the closure to the thread pool instead of a newly spawned thread.

File: wc-rust/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,97 @@
+use std::thread;
```
Comment (dan-f) [inline wc-rust/src/lib.rs]: This is so cool! I've been wondering how thread pools actually work.

**Comment by dan-f** [comment]
This is rad. Excellent work @max-b !!

**Pull request merged**